### ISSUE TITLE
manual_memcpy false positive on overlapping slices

### ISSUE THREAD

**Issue by pornel** [rust-lang-nursery/rust-clippy#2123]
I'm performing operation analogous to `memmove`, not `memcpy`, so I can't replace it with `.clone_from_slice`:

```rust
for i in 0..w {
   inout[d + i] = inout[s + i];
}
```

This copy can't be done with slices, because it'd violate aliasing rules. Suggestion given by clippy does not compile:

> help: try replacing the loop by: `inout[d..(w + d)].clone_from_slice(&inout[s..(w + s)])`

Perhaps clippy should detect that the source and destination are the same?

**Comment by clippered** [comment]
Hi, i would like to give this a try, if nobody is looking at this. Is checking if the variable names are the same enough for this?

**Comment by pornel** [comment]
Thanks for checking it out! Checking for the same expression on both sides will cover most common cases (it could be `struct.field[b+i] = struct.field[a+i]`). 

More complex cases are theoretically possible, but for that you'd probably need a full borrow checker.
